Doc set for VMFR (#110937)
openshift/openshift-docs: modules/oadp-vmfr-creating-vmfr.adoc

// Module included in the following assemblies:
//
// backup_and_restore/application_backup_and_restore/oadp-vmfr/oadp-using-vmfr.adoc

:_mod-docs-content-type: PROCEDURE
[id="oadp-vmfr-creating-vmfr_{context}"]
= Creating a VirtualMachineFileRestore CR

[role="_abstract"]

> Create a `VirtualMachineFileRestore` (VMFR) custom resource (CR) to make files from discovered VM backups accessible for browsing and downloading. This helps you to recover individual files without restoring the entire virtual machine.

Reviewer: ```suggestion
Create a `VirtualMachineFileRestore` (VMFR) custom resource (CR) to make files from discovered virtual machine (VM) backups accessible for browsing and downloading. You can recover individual files without restoring the entire VM.
```

After you create a VMFR CR, the CR undergoes the following phases:

* The initial phase for the CR is `New`.
* The controller validates the referenced discovery, restores PVCs, and creates file-serving resources.
* Upon successful setup, the `status.phase` field of the VMFR CR is updated to `Completed`.

[NOTE]
====
All VMFR custom resources must be created in the protected namespace, which is `openshift-adp` by default.
====

.Prerequisites

* You are logged in to the cluster with the `cluster-admin` role.
* You have installed the {oadp-short} Operator.
* You have configured the `DataProtectionApplication` (DPA) CR with the VMFR feature enabled.
* You have created a `VirtualMachineBackupsDiscovery` (VMBD) CR and its `status.phase` is `Completed`.

.Procedure

> . Create a secret containing the credentials for accessing the file browser:

Reviewer 2: This step is optional

Reviewer: @shdeshpa07 - you do not appear to have marked this as optional

Author: Done. marked this as optional.

+
[source,yaml]
----
apiVersion: v1
kind: Secret
metadata:
  name: vmfr-credentials
  namespace: openshift-adp
type: Opaque
data:
  password: <base64_encoded_password>
  username: <base64_encoded_username>
----
+
where:

`password`:: Specifies the base64-encoded password for the file browser. The password must be at least 12 characters long before encoding.
`username`:: Specifies the base64-encoded username for the file browser.

. Create a `VirtualMachineFileRestore` CR YAML manifest file with the following configuration.
+
To configure file browser access:
+
[source,yaml]
----
apiVersion: oadp.openshift.io/v1alpha1
kind: VirtualMachineFileRestore
metadata:
  name: restore-config-files
  namespace: openshift-adp
spec:
  backupsDiscoveryRef: find-my-vm-backups
  selectedBackups:
    - backup-2025-09-20
    - backup-2025-09-15
  fileAccess:
    fileBrowser:

>       credentialsSecretRef:

Reviewer 2: Optional, controller creates if not specified

Author: done.

metadata:
  name: restore-config-files
  namespace: openshift-adp
spec:
  backupsDiscoveryRef: find-my-vm-backups
  selectedBackups:
    - backup-2025-09-20
    - backup-2025-09-15
  fileAccess:
    ssh: {}
----
+
where:

`name`:: Specifies a name for the VMFR CR. For example, `restore-config-files`.
`namespace`:: Specifies the namespace where the VMFR CR is created. This must be the {oadp-short} protected namespace, typically `openshift-adp`.
`backupsDiscoveryRef`:: Specifies the name of the VMBD CR that contains the discovery results. This VMBD CR must be in the `Completed` phase.
`selectedBackups`:: Specifies a list of backup names from the VMBD valid results to restore. A Velero restore operation is created for each backup in this list.
`fileAccess`:: Specifies the configuration for accessing the restored files.
`fileBrowser`:: Specifies the file browser configuration for web-based access.
`credentialsSecretRef`:: Specifies the name of the Kubernetes secret that contains the credentials for file browser access.
`exposeExternally`:: Specifies whether to create an external route for accessing the file browser. Set to `true` to create a publicly accessible route.

> `ssh`:: Specifies the SSH access configuration. Set to `{}` to enable SSH access with autogenerated credentials. The controller generates an SSH key pair and stores it in a Kubernetes secret.

Reviewer 2: Users can specify the username and publicKey as well. 

```
   ssh: 
     username: fedora
     publicKey: ""
```

Reviewer: @shdeshpa07 - does Prasad's comment need to be added?

Author: @anarnold97 - Yes, I am now going to address all of Prasad's feedback. thanks.

Author: done. added this section.
